### PR TITLE
Display custom FIAMs with SwiftUI

### PR DESCRIPTION
**Pull request by christibbs**
Hey all, opening this PR up early to facilitate API review.

I'll add a test project as well-- for now, see the Code Samples section of the API review document for usage.

### CODE REVIEW

File: FirebaseInAppMessagingSwift.podspec
Comment (paulb777) [inline FirebaseInAppMessagingSwift.podspec:3]: Match the beta naming and versioning of the other Firebase*Swift.podspecs

File: FirebaseInAppMessagingSwift.podspec
Comment (paulb777) [inline FirebaseInAppMessagingSwift.podspec:17]: Change this too

File: FirebaseInAppMessagingSwift.podspec
Comment (christibbs) [reply FirebaseInAppMessagingSwift.podspec]: Fixed in b58a326.

File: FirebaseInAppMessagingSwift.podspec
Comment (christibbs) [reply FirebaseInAppMessagingSwift.podspec]: Fixed in b58a326.

File: FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift
Comment (ryanwilson) [inline FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift:20]: Nit: The `EmptyView()` should be in an else for consistency - it doesn't need to be there when the `if let` succeeds. it doesn't have any effect on the UI but from a code structure at least.

Here and below, as well as the other files in these 2 commits.

File: FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift
Comment (peterfriese) [reply FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift:78]: (This was already addressed by @ryanwilson)

File: FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift
Comment (christibbs) [reply FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift:8]: Just reading about @StateObject now. [Documentation](https://developer.apple.com/documentation/swiftui/stateobject) says it should be used in a View, App, or Scene. Do you think that extends to ViewModifiers? Wdyt @ryanwilson?

File: FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift
Comment (christibbs) [reply FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift:78]: There's an implicit return from 76, but I agree-- this will clean things up a bit.

File: FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift
Comment (christibbs) [reply FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift]: Resolved in dd358fc.

**Comment by ryanwilson** [comment]
Looks like there are some formatting issues in the `check` - can you please fix?

**Comment by christibbs** [comment]
Re-pinging for review now that API review has been approved.

File: FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/Podfile
Comment (paulb777) [inline FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/Podfile:13]: Can this be 13 now?

**Review by paulb777 — COMMENTED** [review]
Please add build and test of the new podspec to CI. (See FirebaseStorageSwift as a reference)

Please add the new library to `Package.swift`

File: FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/Podfile
Comment (christibbs) [reply FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/Podfile]: Fixed in 3e90c8f.

**Comment by christibbs** [comment]
> Please add build and test of the new podspec to CI. (See FirebaseStorageSwift as a reference)
Made an effort at this in f908781, may take some tweaking.

> Please add the new library to `Package.swift`
Fixed in d4d4b0d.

**Comment by paulb777** [comment]
Thanks. I'll wait for green before doing another review pass unless you have any specific questions.

File: FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift
Comment (christibbs) [reply FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift]: Fixed in cfd1bd7.

File: FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift
Comment (peterfriese) [inline FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift:57]: Might not need to explicitly declare the type, as it can be inferred.

File: FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift
Comment (peterfriese) [inline FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift:123]: Same.

File: FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift
Comment (christibbs) [reply FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift]: Removed in 34ba6ac.

File: FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift
Comment (christibbs) [reply FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift]: Removed in 34ba6ac.

**Comment by christibbs** [comment]
This builds, but with unit tests disabled. Going to have to find a way to initialize a mock `InAppMessageDisplayMessage` in the unit tests, given that class's unavailable `init`. (See failures from 0eec2bb.)

File: .github/workflows/inappmessaging.yml
Comment (paulb777) [inline .github/workflows/inappmessaging.yml:46]: Putting the tests in separate jobs either via matrix or just another block makes it easier to debug.

File: FirebaseInAppMessaging/CHANGELOG.md
Comment (paulb777) [inline FirebaseInAppMessaging/CHANGELOG.md:1]: Will now be 7.9.0

File: FirebaseInAppMessaging/Sources/Public/FirebaseInAppMessaging/FIRInAppMessagingRendering.h
Comment (paulb777) [inline FirebaseInAppMessaging/Sources/Public/FirebaseInAppMessaging/FIRInAppMessagingRendering.h:168]: @ryanwilson Is this ok without an API review.

Would it be better to add it to a category inside the test instead of to the public header?

Comment (christibbs) [reply FirebaseInAppMessaging/Sources/Public/FirebaseInAppMessaging/FIRInAppMessagingRendering.h:168]: Can't do it as a category unfortunately, since the tests are in Swift.

Comment (paulb777) [reply FirebaseInAppMessaging/Sources/Public/FirebaseInAppMessaging/FIRInAppMessagingRendering.h:168]: Oh right, of course.

**Review by paulb777 — COMMENTED** [review]
LGTM - deferring approval to @ryanwilson for API approval

File: FirebaseInAppMessaging/Sources/Public/FirebaseInAppMessaging/FIRInAppMessagingRendering.h
Comment (ryanwilson) [reply FirebaseInAppMessaging/Sources/Public/FirebaseInAppMessaging/FIRInAppMessagingRendering.h:168]: Yeah, this needs an API review. Do you mind setting that up?

File: FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift
Comment (ryanwilson) [inline FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift:81]: I think it'd be a bit more clear if we keep the `self.` before modifier

File: FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift
Comment (ryanwilson) [inline FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift:147]: Same here re: `self`

File: FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift
Comment (ryanwilson) [inline FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift:154]: Nit: can you please add documentation or this since it's the most critical class of the bunch here?

File: FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/Shared/ContentView.swift
Comment (ryanwilson) [inline FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/Shared/ContentView.swift:19]: Nit: `@State` variables should also be marked with `private` so they aren't modified outside of this scope.

File: FirebaseInAppMessaging/Swift/Tests/Unit/CustomInAppMessageDisplayViewModifierTests.swift
Comment (ryanwilson) [inline FirebaseInAppMessaging/Swift/Tests/Unit/CustomInAppMessageDisplayViewModifierTests.swift:31]: Can we verify anything on the  `TestDelegate()`  here?

File: FirebaseInAppMessagingSwift.podspec
Comment (ryanwilson) [inline FirebaseInAppMessagingSwift.podspec:3]: This should be `7.9.0-beta` now

File: FirebaseInAppMessagingSwift.podspec
Comment (ryanwilson) [inline FirebaseInAppMessagingSwift.podspec:4]: In`for Firebase In-App Messaging`

File: FirebaseInAppMessagingSwift.podspec
Comment (ryanwilson) [inline FirebaseInAppMessagingSwift.podspec:24]: This  line can be removed.

File: FirebaseInAppMessagingSwift.podspec
Comment (ryanwilson) [inline FirebaseInAppMessagingSwift.podspec:27]: This line can be removed.

File: Package.swift
Comment (ryanwilson) [inline Package.swift:769]: You may need to do something like `AppDistribution` and add in a condition. Also, can you add a line to make sure this compiles in the swift-test as well?

**Review by ryanwilson — COMMENTED** [review]
Just the API review and a few small nits but otherwise LGTM!

File: FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift
Comment (christibbs) [reply FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift]: Thanks. Changed in 99b4a7f.

File: FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift
Comment (christibbs) [reply FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift:147]: Fixed in 99b4a7f.

File: FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift
Comment (christibbs) [reply FirebaseInAppMessaging/Swift/Source/CustomInAppMessageDisplayViewModifier.swift:158]: Fixed in 99b4a7f.

File: FirebaseInAppMessagingSwift.podspec
Comment (christibbs) [reply FirebaseInAppMessagingSwift.podspec]: Fixed in fae7aa2.

File: FirebaseInAppMessagingSwift.podspec
Comment (christibbs) [reply FirebaseInAppMessagingSwift.podspec]: Fixed in fae7aa2.

File: FirebaseInAppMessagingSwift.podspec
Comment (christibbs) [reply FirebaseInAppMessagingSwift.podspec]: Fixed in fae7aa2.

File: Package.swift
Comment (christibbs) [reply Package.swift]: First part fixed in fae7aa2.

@ryanwilson I'm unclear about what you mean by the line to ensure it compiles in swift-test.

File: FirebaseInAppMessaging/Sources/Public/FirebaseInAppMessaging/FIRInAppMessagingRendering.h
Comment (christibbs) [reply FirebaseInAppMessaging/Sources/Public/FirebaseInAppMessaging/FIRInAppMessagingRendering.h:168]: Kicked off API review process, will track internally.

File: Package.swift
Comment (ryanwilson) [inline Package.swift:770]: tvOS too :)

Comment (christibbs) [reply Package.swift]: Fixed in c5fcc43.

File: Package.swift
Comment (christibbs) [reply Package.swift]: Added tests in c5fcc43.

File: FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/Podfile
Comment (paulb777) [inline FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/Podfile:1]: Ideally there should be a build test in CI for the Integration test project, but should be ok to do later.

File: FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/Shared/ContentView.swift
Comment (paulb777) [reply FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/Shared/ContentView.swift:19]: Still open

File: FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/Podfile
Comment (paulb777) [inline FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/Podfile:1]: An initial first line should be inserted:

`source 'https://github.com/firebase/SpecsDev.git'`

See other Podfiles in the repo for an example

File: SwiftPMTests/swift-test/main.swift
Comment (paulb777) [inline SwiftPMTests/swift-test/main.swift:72]: Nice!

**Review by paulb777 — APPROVED** [review]

File: FirebaseInAppMessaging/Swift/Tests/Unit/CustomInAppMessageDisplayViewModifierTests.swift
Comment (christibbs) [reply FirebaseInAppMessaging/Swift/Tests/Unit/CustomInAppMessageDisplayViewModifierTests.swift:31]: @ryanwilson The only thing we're missing is testing that the impression fires in `onAppear()` see line 34 of CustomInAppMessageDisplayViewModifier.swift for example. I don't see a way of doing this without opening up the initializers for the other message subclasses (since it involves mocking the entire flow). Let's table this and handle it when we handle UI testing for the integration test app.

File: FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/Podfile
Comment (christibbs) [reply FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/Podfile:2]: Resolved in 5d8b009.

File: FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/Shared/ContentView.swift
Comment (christibbs) [reply FirebaseInAppMessaging/Swift/Tests/Integration/FIAMSwiftUI/Shared/ContentView.swift]: Fixed in 5d8b009.

**Review by ryanwilson — APPROVED** [review]
Thanks for working through the comments!

**Pull request merged**
